mesonbuild/scripts/clippy.py

```diff
@@ -0,0 +1,66 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2024 The Meson development team
+
+from __future__ import annotations
+from collections import defaultdict
+import os
+import tempfile
+import typing as T
+
+from .run_tool import run_tool_on_targets, run_with_buffered_output
+from .. import build, mlog
+from ..compilers.rust import RustCompiler
+from ..mesonlib import MachineChoice, PerMachine
+
+class ClippyDriver:
+    def __init__(self, build: build.Build, tempdir: str):
+        self.tools: PerMachine[T.List[str]] = PerMachine([], [])
+        self.warned: T.DefaultDict[str, bool] = defaultdict(lambda: False)
+        self.tempdir = tempdir
+        for machine in MachineChoice:
+            compilers = build.environment.coredata.compilers[machine]
+            if 'rust' in compilers:
+                compiler = compilers['rust']
+                assert isinstance(compiler, RustCompiler)
```

Comment:
This assert is the only reason we're importing compilers.rust, but the reason for the assert seems to be solely for type checking.

The assert pulls in a runtime dependency but allows using python -OO to optimize it out to nothing. The cost/benefit ratio here isn't great. We can use `T.cast('RustCompiler', compilers['rust'])` instead, which allows moving the import to T.TYPE_CHECKING